### PR TITLE
feat!: Token Registry V4

### PR DESCRIPTION
**Pull request by osslgtm**
## Summary
Updated Token-Registry from V2 to V4

### Removed Features:
Title Escrow deploy (Breaking!)

### Updated Smart Contract Functions (Breaking!):
Deploy
Issue
Transfer - Nominate, Endorse, Transfer
Surrender - Accept, Reject
...

Excluded features from Token Registry 
Deploy - doesn't have standalone deployment and verify flags

## Relevant PRs

https://github.com/Open-Attestation/open-attestation-cli/pull/212
https://github.com/Open-Attestation/open-attestation-cli/pull/220
https://github.com/Open-Attestation/open-attestation-cli/pull/222
https://github.com/Open-Attestation/open-attestation-cli/pull/223
https://github.com/Open-Attestation/open-attestation-cli/pull/224

https://github.com/Open-Attestation/open-attestation-cli/pull/207

https://github.com/Open-Attestation/open-attestation-cli/pull/228
https://github.com/TradeTrust/tradetrust-website/pull/606
https://github.com/TradeTrust/document-creator-website/pull/254

### CODE REVIEW

File: src/implementations/deploy/token-registry/token-registry.ts
Comment (cavacado) [inline src/implementations/deploy/token-registry/token-registry.ts:107]: move to token-registry/helpers file if possible and unit test them if necessary

**Review by cavacado — APPROVED** [review]
im approving, try to get another pair of eyes?

File: src/implementations/title-escrow/transferOwner.test.ts
Comment (isaackps) [inline src/implementations/title-escrow/transferOwner.test.ts:5]: should this be change to beneficiary? file name as well.

File: src/implementations/title-escrow/transferOwner.ts
Comment (isaackps) [inline src/implementations/title-escrow/transferOwner.ts:21]: In this file and the file name, if its only beneficiary should change to beneficiary, if is for transferring of beneficiary and holder should be Owners right??

File: README.md
Comment (isaackps) [inline README.md:600]: this too, can change to newBeneficiary? it will be much easier to understand right?

File: src/commands/title-escrow/endorse-transfer-of-owner.ts
Comment (isaackps) [inline src/commands/title-escrow/endorse-transfer-of-owner.ts:31]: same here k

File: src/commands/title-escrow/title-escrow-command.type.ts
Comment (isaackps) [inline src/commands/title-escrow/title-escrow-command.type.ts:14]: same here

File: src/implementations/title-escrow/transferOwners.ts
Comment (isaackps) [inline src/implementations/title-escrow/transferOwners.ts:24]: plural Owners

**Review by isaackps — COMMENTED** [review]
i might have miss out some place too, but can u update them? so that we wont get confuse next time. thanks!

**Comment by superical** [comment]
What's the plan for the "E2E" tests you were proposing during the call last week? I don't see them in this new PR anymore. Did you manage to do something to them to improve the tests or you left them out of this PR?

File: README.md
Comment (simboonlong) [inline README.md:250]: is it correct to say without `--factory-address`, it fallbacks to [these](https://github.com/Open-Attestation/token-registry/blob/70ff40eea7406866fb595147522a964c8d78f3b6/src/constants/contract-address.ts#L10-L16)?

File: src/commands/deploy/deploy.types.ts
Comment (simboonlong) [inline src/commands/deploy/deploy.types.ts:19]: (qn) i see that these will fallback to the hardcoded values as well, can help me understand the use of these other 2 params?

File: src/implementations/deploy/token-registry/helpers.test.ts
Comment (simboonlong) [inline src/implementations/deploy/token-registry/helpers.test.ts:34]: use expect toThrow?

File: src/implementations/deploy/token-registry/helpers.test.ts
Comment (simboonlong) [inline src/implementations/deploy/token-registry/helpers.test.ts:47]: what was being tested seems to be ethers utils method, i not sure if this is necessary but ok to leave it

File: src/implementations/deploy/token-registry/token-registry.test.ts
Comment (simboonlong) [inline src/implementations/deploy/token-registry/token-registry.test.ts:77]: why was this removed?

File: src/implementations/deploy/token-registry/token-registry.test.ts
Comment (simboonlong) [inline src/implementations/deploy/token-registry/token-registry.test.ts:20]: describe describe?

File: src/implementations/document-store/issue.test.ts
Comment (simboonlong) [inline src/implementations/document-store/issue.test.ts:66]: why is this removed? we no longer accepting `key-file` method?

File: src/implementations/document-store/revoke.test.ts
Comment (simboonlong) [inline src/implementations/document-store/revoke.test.ts:51]: same 🤔

File: src/implementations/document-store/transfer-ownership.test.ts
Comment (simboonlong) [inline src/implementations/document-store/transfer-ownership.test.ts:51]: same, starting to sus myself now

File: src/implementations/title-escrow/rejectSurrendered.ts
Comment (simboonlong) [inline src/implementations/title-escrow/rejectSurrendered.ts:15]: looks like 👍, i would presume these are handled at token registry end via `restore` now?

File: src/implementations/title-escrow/transferOwners.ts
Comment (simboonlong) [inline src/implementations/title-escrow/transferOwners.ts:4]: why needa alias? how about rename directly?

File: src/implementations/token-registry/issue.test.ts
Comment (simboonlong) [inline src/implementations/token-registry/issue.test.ts:35]: (nitpick): the casing caught my eye, thinking of camelcase regardless of acronym / abbreviation. ok to leave as this too

File: src/implementations/document-store/issue.test.ts
Comment (simboonlong) [reply src/implementations/document-store/issue.test.ts:66]: ah ic, then its fine no need to restore

**Review by simboonlong — APPROVED** [review]
do rectify the linter

**Review by superical — APPROVED** [review]
In case there are anything that could have been missed due to the massive changes, I think it will make be good to continue to look for any issues even after merging. Thanks for the efforts!

**Pull request merged**
